application/classes/Ushahidi/Formatter/Post/CSV.php

```diff
@@ -34,25 +35,44 @@ class Ushahidi_Formatter_Post_CSV extends Ushahidi_Formatter_API
 	protected $search;
 	protected $fs;
 	protected $tmpfname;
-	protected $add_header;
+	protected $add_header = true;
+	protected $heading;
 
 	// Formatter
-	public function __invoke($records)
+	public function __invoke($records, $attributes = [])
 	{
-		return $this->generateCSVRecords($records);
+		if ($this->heading) {
+			return $this->generateCSVRecords($records, $attributes);
+		} else {
+			//throw exception
+		}
 	}
 
 	public function setAddHeader($add_header)
 	{
 		$this->add_header = $add_header;
 	}
 
+
+	public function setHeading($heading)
+	{
+		$this->heading = $heading;
+	}
+
 	public function setFilesystem($fs)
 	{
 		$this->tmpfname = "tmp" . DIRECTORY_SEPARATOR . strtolower(uniqid() . '-' . strftime('%G-%m-%d') . '.csv');
 		$this->fs = $fs;
 	}
 
+	public function createHeading($attributes, $records)
+	{
```

Comment:
maybe $attributes can be set as a property of the formatter to avoid passing it around?

Reply:
it's not the same as heading because it has all the data to decide what to do about each field